walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,135 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
+  /*
+   * Assigns attendees from request to local HashMap to search for them in constant time.
+   */
+  private HashSet<String> toSet(MeetingRequest request) {
+    HashSet<String> requestAttendees = new HashSet<String>();
+    for(String attendee : request.getAttendees()) {
+      requestAttendees.add(attendee);
+    }
+    return requestAttendees;
+  }
+
+  /*
+   * Adds a TimeRange depending on the scenario. Some additions use a recursive call to this same
+   * function because if a new TimeRange is formed from a merge or an existing TimeRange is replaced
+   * because of a containment issue, it needs to be checked against all the previously added
+   * TimeRanges for new overlap or containment issues.
+   */
+  private ArrayList<TimeRange> addTimeRange(ArrayList<TimeRange> occupiedTimeSlots,
+      TimeRange addedTimeSlot) {
+    for(TimeRange timeSlot : occupiedTimeSlots) {
+      // New TimeRange is already contained, so it is not added.
+      if(timeSlot.contains(addedTimeSlot)) {
+        return occupiedTimeSlots;
+      }
+      // New TimeRange contains an already existing meeting, so the new replaces the existing.
+      if(addedTimeSlot.contains(timeSlot)) {
+        occupiedTimeSlots.remove(timeSlot);
+        addTimeRange(occupiedTimeSlots, addedTimeSlot);
+        return occupiedTimeSlots;
+      }
+      // New TimeRange overlaps with an already existing meeting.
+      if(timeSlot.overlaps(addedTimeSlot)) {
+        addTimeRange(occupiedTimeSlots, mergeTimeRanges(timeSlot, addedTimeSlot));
+        return occupiedTimeSlots;
+      }
+    }
+    // No scenario applied to TimeRange added normally.
+    occupiedTimeSlots.add(addedTimeSlot);
+    return occupiedTimeSlots;
+  }
+
+  /*
+   * Merges overlapping TimeRanges into a single TimeRange that includes both of them.
+   */
+  private TimeRange mergeTimeRanges(TimeRange a, TimeRange b) {
+    return TimeRange.fromStartEnd(a.start() < b.start() ? a.start() : b.start(),
+        a.end() > b.end() ? a.end() : b.end(), false);
+  }
+
+  /*
+   * Returns a collection with TimeRanges that are not available for meetings.
+   */
+  private ArrayList<TimeRange> getOccupiedTimeSlots(Collection<Event> events,
+      HashSet<String> requestAttendees) {
+    ArrayList<TimeRange> timeRangeEventsRequestAttendees = new ArrayList<TimeRange>();
```

Comment:
This name makes me feel like you are maintaining an ArrayList of RequestAttendees, maybe change the name so it's more clear that this is an array of unavailable TimeRanges for the requested attendees?

Reply:
Ack!